src/components/external-source/ExternalSourceManager.svelte

```diff
@@ -0,0 +1,904 @@
+        }
+      },
+    },
+  ];
+
+  let dataGrid: DataGrid<ExternalSourceSlim> | undefined = undefined;
+
+  let keyInputField: HTMLInputElement; // need this to set a focus on it. not related to the value
+
+  let keyField = field<string>('', [required]);
+  let sourceTypeField = field<string>('', [required]); // need function to check if in list of allowable types...
+  let derivationGroupField = field<string>('', [required]);
+  let startTimeDoyField = field<string>('', [required, timestamp]); // requires validation function
+  let endTimeDoyField = field<string>('', [required, timestamp]); // requires validation function
+  let validAtDoyField = field<string>('', [required, timestamp]); // requires validation function
+
+  // table variables
+  let columnDefs: DataGridColumnDef[] = baseColumnDefs;
+
+  // for external events table
+  let externalEventsTableFilterString: string = '';
+
+  // source detail variables
+  let selectedSource: ExternalSourceSlim | null = null;
+  let selectedSourceId: string | null = null;
+
+  // Selected element variables
+  let selectedEvent: ExternalEvent | null = null;
+  let selectedRowId: ExternalEventId | null = null;
+  let selectedEvents: ExternalEvent[] = [];
+
+  // We want to parse a file selected for upload.
+  let files: FileList | undefined;
+  let file: File | undefined;
+  let parsedExternalSource: ExternalSourceJson | undefined;
+
+  // For filtering purposes (modelled after TimelineEditorLayerFilter):
+  let filterExpression: string = '';
+
+  // External source + derivation group creation variables
+  let selectedSourceLinkedDerivationGroupsPlans: PlanDerivationGroup[] = [];
+
+  // Permissions
+  let hasDeleteExternalSourcePermissionOnSelectedSource: boolean = false;
+  let hasCreatePermission: boolean = false;
+
+  let isDerivationGroupFieldDisabled: boolean = true;
+
+  let gridRowSizes: string = '1fr 3px 0fr';
+
+  // Clear all error stores when a source is selected as they will not be shown
+  $: if (selectedSource !== null) {
+    createExternalSourceError.set(null);
+    createExternalSourceTypeError.set(null);
+    createDerivationGroupError.set(null);
+    parsingError.set(null);
+  }
+
+  $: if (selectedSource !== null) {
+    hasDeleteExternalSourcePermissionOnSelectedSource = featurePermissions.externalSource.canDelete(user, [
+      selectedSource,
+    ]);
+  }
+
+  $: selectedSourceId = selectedSource
+    ? getExternalSourceRowId({ derivation_group_name: selectedSource.derivation_group_name, key: selectedSource.key })
+    : null;
+
+  // File parse logic
+  $: if (files) {
+    // Safeguard against infinitely executing parse logic
+    if (file !== files[0]) {
+      createExternalSourceError.set(null);
+      createExternalSourceTypeError.set(null);
+      createDerivationGroupError.set(null);
+      parsingError.set(null);
+      isDerivationGroupFieldDisabled = true;
+
+      file = files[0];
+      if (/\.json$/.test(file.name)) {
```

Comment:
If you select an invalid json file, then open the selection window again and hit escape without selecting a file you get this uncaught error in the console:
`Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'name')` so think we need to actually check that the file exists somewhere in this logic.

Reply:
Which causes other functionality on the page to break

Reply:
Weirdly this doesn't seem to happen in Firefox, but I was able to replicate in Chrome. Seems to be fixed by adding a `!== undefined` check with the filename test!